doc/migration.md

```diff
@@ -0,0 +1,424 @@
+* `sf::Image::saveToMemory` returns a `std::optional` because the `sf::Image` may be empty or the underlying implementation may fail.
+* `sf::SoundFileReader::open` returns a `std::optional` because the stream being opened may not be valid.
+* `sf::Music::onLoop` and `sf::SoundStream::onLoop` returns a `std::optional` because if the objects are not in a looping state then there is nothing to return.
+* `sf::InputStream` uses `std::optional` in various places.
+Instead of returning `-1` to signal an error, `std::nullopt` can be returned.
+
+LearnCpp.com is a great place to learn more about using `std::optional`.
+Read more about that [here](https://www.learncpp.com/cpp-tutorial/stdoptional/).
+
+## Fixed Width Integers
+
+SFML 2 contained various typedefs for fixed width integers.
+Those are now replaced with the fixed width integers provided in the `<cstdint>` header from the standard library.
+
+| v2           | v3              |
+| ------------ | --------------- |
+| `sf::Int8`   | `std::int8_t`   |
+| `sf::Uint8`  | `std::uint8_t`  |
+| `sf::Int16`  | `std::int16_t`  |
+| `sf::Uint16` | `std::uint16_t` |
+| `sf::Int32`  | `std::int32_t`  |
+| `sf::Uint32` | `std::uint32_t` |
+| `sf::Int64`  | `std::int64_t`  |
+| `sf::Uint64` | `std::uint64_t` |
+
+## New Constructors for Loading Resources
+
+The following classes gained constructors that allow for loading/opening resources in a single expression.
+Upon failure they throw an `sf::Exception`.
+
+* `sf::InputSoundFile`
+* `sf::OutputSoundFile`
+* `sf::Music`
+* `sf::SoundBuffer`
+* `sf::Font`
+* `sf::Image`
+* `sf::RenderTexture`
+* `sf::Shader`
+* `sf::Texture`
+* `sf::FileInputStream`
+* `sf::Cursor`
+
+SFML 3 still supports the SFML 2 style of error handling in addition to these new constructors.
+
+v2:
+```cpp
+sf::SoundBuffer soundBuffer;
+if (!soundBuffer.loadFromFile("sound.wav"))
+{
+    // Handle error
+}
+```
+
+v3:
+```cpp
+sf::SoundBuffer soundBuffer;
+if (!soundBuffer.loadFromFile("sound.wav"))
+{
+    // Handle error
+}
+
+// OR
+
+const sf::SoundBuffer soundBuffer("sound.wav");
+```
+
+## `sf::Vector2<T>` and `sf::Vector3<T>` Utility Functions
+
+`sf::Vector2<T>` and `sf::Vector3<T>` gained a number of new functions for performing common mathematic operations on vectors.
+
+| `sf::Vector2<T>` Function    | Description                              |
+| ---------------------------- | ---------------------------------------- |
+| `Vector2(T, sf::Angle)`      | Construct from polar coordinates         |
+| `length()`                   | Get length                               |
+| `lengthSq()`                 | Get length squared                       |
+| `normalized()`               | Get vector normalized to unit circle     |
+| `angleTo(sf::Vector2)`       | Get angle to another vector              |
+| `angle()`                    | Get angle from origin                    |
```

Comment:
Origin means (0, 0) to me. This should be phrased as angle from x axis or similar.